### PR TITLE
[unifiedremote] Initial contribution

### PR DESCRIPTION
**Pull request by GiviMAD**
Copied from initial pr(#8494):
## Unified Remote Binding.
This binding allow send some command to the unified remote server through its web client interface.
It create some channels to control the mouse, some navigation keys, some media keys, and not much more.

_Good day.  This is my first contribution to this amazing product. Some days ago I started search for a way to control my computer with openHAB and I decided to try to write this binding that seems to finally be working._
_Please let me know if there are any suggestions to improve the binding as I'm still getting in touch with this technology. I will also open a thread in the Add-ons forum pointing to this pr, in case someone wants to try it or have an idea for a better implementation. Thanks in advance for your time._

[download binding jar](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/addons/bundles/org.openhab.binding.unifiedremote/3.0.0-SNAPSHOT/org.openhab.binding.unifiedremote-3.0.0-SNAPSHOT.jar)

### CODE REVIEW

File: bundles/org.openhab.binding.unifiedremote/.classpath [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/.classpath:1]: This file is generated automatically by maven and should not be part of the PR.

File: bundles/org.openhab.binding.unifiedremote/.project [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/.project:1]: This file is generated automatically by maven and should not be part of the PR.

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+This binding integrates the [Unified Remote Server](https://www.unifiedremote.com/).
+
+<b>Known Limitations: It needs the web interface to be enabled on the server settings to work.</b>
+
+## Discovery
+
+Discovery works on the default discovery UDP port 9511.
+
+## Thing Configuration
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:11]: In this chapter, the Thing Type ID should be mentioned.

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+The Unified Remote Server Thing requires the host to be correctly configured in order to work correctly.
+Other properties like tcpPort and udpPort are not used in the initial implementation.
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:14]: This sentence could be removed and re-added, when it's implemented.

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+```
+Thing unifiedremote:server:xx-xx-xx-xx-xx-xx [ host="192.168.1.10" ]
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:17]: This example should be moved to the end to the chapter "Full Example". You could also add an example of the .items file and the .sitemap file (which are more important that the .things file).

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+<b>Known Limitations: It needs the web interface to be enabled on the server settings to work.</b>
+
+## Discovery
+
+Discovery works on the default discovery UDP port 9511.
+
+## Thing Configuration
+
+The Unified Remote Server Thing requires the host to be correctly configured in order to work correctly.
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:13]: Can you mention the parameter name of the address?

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+```
+Thing unifiedremote:server:xx-xx-xx-xx-xx-xx [ host="192.168.1.10" ]
+```
+
+## Channels
+
+_Here you should provide information about available channel types, what their meaning is and how they can be used._
+
+_Note that it is planned to generate some part of this based on the XML files within ```src/main/resources/OH-INF/thing``` of your binding._
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:24]: These are only placeholders.

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+| channel  | type   | description                  |
+|----------|--------|------------------------------|
+| mouse-move  | String | Mouse Move. Expect number JSON array ("[10,10]").   |
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:28]: Are these absolute or relative coordinates? Is this in pixels?

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+/**
+ * The {@link UnifiedRemoteConnection} Handles Remote Server Communications
+ *
+ * @author Miguel Alvarez - Initial contribution
+ */
+@NonNullByDefault
+public class UnifiedRemoteConnection implements Closeable {
+    private final Logger logger = LoggerFactory.getLogger(UnifiedRemoteConnection.class);
+    private final String url;
+    private final HttpClient http = new HttpClient();
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java:44]: The HTTP can be retrieved from the framework via injection. Then, resources are re-used. See for example https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.automower/src/main/java/org/openhab/binding/automower/internal/AutomowerHandlerFactory.java#L60

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    private final JsonParser jsonParser = new JsonParser();
+    private final int WEB_CLIENT_PORT = 9510;
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java:46]: This could be static. Same for the upper case fields below. Static fields should be moved to the top of the class.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    public UnifiedRemoteConnection(String host) {
+        url = "http://" + host + ":" + WEB_CLIENT_PORT + "/client/";
+    }
+
+    public void authenticate() throws Exception {
+        ContentResponse response = null;
+        connectionGUID = "web-" + UUID.randomUUID().toString();
+        if (!http.isStarted()) {
+            http.start();
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java:66]: This could be removed if you use the framework's HttpClient.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+        response = http.GET(getPath("connect"));
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java:67]: Better specify a timeout. Many bindings use 10 sec.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+
+    @Override
+    public void close() {
+        if (http.isStarted()) {
+            try {
+                http.stop();
+            } catch (Exception e) {
+            }
+        }
+    }
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteConnection.java:271]: This can also be removed if using the framework's HttpClient.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+ *
+ * @author Miguel Alvarez - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.unifiedremote")
+@NonNullByDefault
+public class UnifiedRemoteDiscoveryService extends AbstractDiscoveryService {
+
+    private static final Logger logger = LoggerFactory.getLogger(UnifiedRemoteDiscoveryService.class);
+    static final int TIMEOUT_MS = 20000;
+    private static final long DISCOVERY_RESULT_TTL = TimeUnit.MINUTES.toSeconds(5);
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:46]: It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+/**
+ * The {@link UnifiedRemoteDiscoveryService} discover Unified Remote Server Instances in the network.
+ *
+ * @author Miguel Alvarez - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.unifiedremote")
+@NonNullByDefault
+public class UnifiedRemoteDiscoveryService extends AbstractDiscoveryService {
+
+    private static final Logger logger = LoggerFactory.getLogger(UnifiedRemoteDiscoveryService.class);
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:44]: Loggers should not be final.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+    private static final long DISCOVERY_RESULT_TTL = TimeUnit.MINUTES.toSeconds(5);
+
+    public UnifiedRemoteDiscoveryService() {
+        super(SUPPORTED_THING_TYPES, TIMEOUT_MS, false);
+    }
+
+    @Override
+    protected void startScan() {
+        UnifiedRemoteUdpDiscovery client = new UnifiedRemoteUdpDiscovery();
+        client.call(serverInfo -> addNewServer(serverInfo));
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:55]: Syntactical sugar
```suggestion
        client.call(this::addNewServer);
```

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    private void addNewServer(UnifiedRemoteUdpDiscovery.ServerInfo serverInfo) {
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(PARAMETER_HOSTNAME, serverInfo.host);
+        properties.put(PARAMETER_TCP_PORT, serverInfo.tcpPort);
+        properties.put(PARAMETER_UDP_PORT, serverInfo.udpPort);
+        thingDiscovered(
+                DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_UNIFIED_REMOTE_SERVER, serverInfo.macAddress))
+                        .withTTL(DISCOVERY_RESULT_TTL).withProperties(properties).withLabel(serverInfo.name).build());
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:65]: Can you add the representationProperty?

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+        /**
+         * String used to replace non printable characters on service response
+         */
+        public static final String NON_PRINTABLE_CHARTS_REPLACEMENT = ": :";
+
+        private static final int MAX_PACKET_SIZE = 2048;
+        /**
+         * maximum time to wait for a reply, in milliseconds.
+         */
+        private static final int TIMEOUT = 3000; // milliseonds
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:90]: See above. Append unit.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+         */
+        public void call(Consumer<ServerInfo> listener) {
+            byte[] receiveBuffer = new byte[MAX_PACKET_SIZE];
+            DatagramPacket receivePacket = new DatagramPacket(receiveBuffer, receiveBuffer.length);
+
+            DatagramSocket socket = null;
+            try {
+                socket = createSocket();
+            } catch (SocketException e) {
+                logger.error("Error creating socket: {}", e.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:157]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+                socket.send(packet);
+                logger.debug("Sent packet to {}:{}", broadcastAddress.getHostAddress(), servicePort);
+                for (int i = 0; i < 20; i++) {
+                    socket.receive(receivePacket);
+                    String host = receivePacket.getAddress().getHostAddress();
+                    logger.debug("Received reply from {}", host);
+                    try {
+                        ServerInfo serverInfo = tryParseServerDiscovery(receivePacket);
+                        listener.accept(serverInfo);
+                    } catch (Exception ex) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:175]: You should specify the concrete exception you want to catch, since catching `Exception` catches also all runtime exceptions.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+                        logger.error("Exception parsing server discovery response from {}: {}", host, ex.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:176]: See above. logging

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+                    }
+                }
+            } catch (SocketTimeoutException ste) {
+                logger.debug("SocketTimeoutException during socket operation: {}", ste.getMessage());
+            } catch (IOException ioe) {
+                logger.error("IOException during socket operation: {}", ioe.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:182]: See above

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+            }
+            socket.close();
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:184]: You could add this to the finally block, to close the socket also if an unchecked exception is thrown.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+    public UnifiedRemoteHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            // TODO: nothing to do for now
+            return;
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:59]: As the Thing doesn't provide any data, this could be removed entirely.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+                }
+                if (isErrorResponse(response)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Session expired");
+                    connection.authenticate();
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Connection not initialized");
+            }
+        } catch (Exception e) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:85]: When using the framework's HttpClient, this could be the concrete Exceptions.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+                ThingStatus status = thing.getStatus();
+                if ((status == ThingStatus.OFFLINE || status == ThingStatus.UNKNOWN) && connection != null) {
+                    connection.authenticate();
+                    updateStatus(ThingStatus.ONLINE);
+                } else if (status == ThingStatus.ONLINE) {
+                    if (isErrorResponse(connection.keepAlive())) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Keep alive failed");
+                    }
+                }
+            } catch (Exception e) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:116]: See above

File: bundles/org.openhab.binding.unifiedremote/src/main/resources/OH-INF/binding/binding.xml [hunk 27]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="unifiedremote" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:binding="https://openhab.org/schemas/binding/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/resources/OH-INF/binding/binding.xml:6]: Please remove the extra empty line.

File: bundles/org.openhab.binding.unifiedremote/src/main/resources/OH-INF/thing/thing-types.xml [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+		<description>Unified Remote Server Thing for Unified Remote Binding</description>
+		<channels>
+			<channel id="mouse-move" typeId="mouse-move-channel"/>
+			<channel id="send-key" typeId="send-key-channel"/>
+		</channels>
+
+		<config-description>
+			<parameter name="host" type="text" required="true">
+				<label>Hostname</label>
+				<description>Unified Remote Server Hostname</description>
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/resources/OH-INF/thing/thing-types.xml:18]: You could specify `<context>network-address</context>` to get a free validation.

**Review by fwolter — CHANGES_REQUESTED** [review]
Thanks for your contribution! I reviewed your code and found only minor things. Good code in general.

File: bundles/org.openhab.binding.unifiedremote/README.md [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+
+<b>Known Limitations: It needs the web interface to be enabled on the server settings to work.</b>
+
+## Discovery
+
+Discovery works on the default discovery UDP port 9511.
+
+## Thing Configuration
+
+Only supported thing is 'Unified Remote Server Thing' witch requires the Hostname to be correctly configured in order to work.
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/README.md:13]:
```suggestion
Only supported thing is 'Unified Remote Server Thing' which requires the Hostname to be correctly configured in order to work.
```

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private void addNewServer(UnifiedRemoteUdpDiscovery.ServerInfo serverInfo) {
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(PARAMETER_HOSTNAME, serverInfo.host);
+        properties.put(PARAMETER_TCP_PORT, serverInfo.tcpPort);
+        properties.put(PARAMETER_UDP_PORT, serverInfo.udpPort);
+        thingDiscovered(
+                DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_UNIFIED_REMOTE_SERVER, serverInfo.macAddress))
+                        .withTTL(DISCOVERY_RESULT_TTL_SEC).withRepresentationProperty(serverInfo.macAddress)
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:65]: The representation property must be the name of the property, not the value.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "Unexpected exception: " + e.getMessage());
+                }
+            }
+        }, 0, 40, TimeUnit.SECONDS);
+    }
+
+    private void stopConnectionChecker() {
+        if (connectionCheckerSchedule != null && !connectionCheckerSchedule.isCancelled()) {
```
Comment (fwolter) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:135]: No harm in canceling a task that is already cancelled.
```suggestion
        if (connectionCheckerSchedule != null) {
```

**Review by fwolter — CHANGES_REQUESTED** [review]
There are some markdown warnings left. You could take a look at `target/code-analysis/report.html`.

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by fwolter** [comment]
Can you fix the conflicting CODEOWNERS file?

**Comment by GiviMAD** [comment]
Done!

**Comment by cpmeister** [comment]
Can you fix the conflicting pom.xml file?

**Comment by GiviMAD** [comment]
the build finally works!

**Comment by cpmeister** [comment]
Almost done, just resolve these warnings and you are finished!
```
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[70,36] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[73,36] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[80,21] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[106,61] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[115,21] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[118,41] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteHandler.java:[136,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:[100,20] The value of the field org.openhab.binding.unifiedremote.internal.UnifiedRemoteDiscoveryService.UnifiedRemoteUdpDiscovery.ServerInfo.publicIp is not used
```

To deal with the null warnings you should assign the field to a local variable and perform your operations on that local variable instead.

**Comment by GiviMAD** [comment]
I think I get rid of all the warnings. Please also note I make a small change in the catch clause in the fie UnifiedRemoteHandler.java to consider the thing as offline is the TimeoutException is thrown. And thanks for the tip on how to deal with the nullable type warnings,I wasn't sure about how to solve those.

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 32]
```diff
@@ -0,0 +1,187 @@
...
+        properties.put(PARAMETER_HOSTNAME, serverInfo.host);
+        properties.put(PARAMETER_TCP_PORT, serverInfo.tcpPort);
+        properties.put(PARAMETER_UDP_PORT, serverInfo.udpPort);
+        thingDiscovered(
+                DiscoveryResultBuilder.create(new ThingUID(THING_TYPE_UNIFIED_REMOTE_SERVER, serverInfo.macAddress))
+                        .withTTL(DISCOVERY_RESULT_TTL_SEC).withRepresentationProperty(PARAMETER_MAC_ADDRESS)
+                        .withProperties(properties).withLabel(serverInfo.name).build());
+    }
+
+    private class UnifiedRemoteUdpDiscovery {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:70]: Is this nested class even needed? It doesn't have any fields. Can you just move all this logic into the UnifiedRemoteDiscoveryService class instead?

File: bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java [hunk 33]
```diff
@@ -0,0 +1,187 @@
...
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UnifiedRemoteDiscoveryService} discover Unified Remote Server Instances in the network.
+ *
+ * @author Miguel Alvarez - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.unifiedremote")
```
Comment (cpmeister) [inline bundles/org.openhab.binding.unifiedremote/src/main/java/org/openhab/binding/unifiedremote/internal/UnifiedRemoteDiscoveryService.java:40]:
```suggestion
@Component(service = DiscoveryService.class, configurationPid = "discovery.unifiedremote")
```

**Review by cpmeister — CHANGES_REQUESTED** [review]
I thought I was done before, but I took another look and found these two things.
These will be my last change suggestions through so fix these and I'll approve and merge this.
Thanks for all your work!

**Comment by GiviMAD** [comment]
Done! Please note I change the exception control in the UnifiedRemoteHandler.java again, I was testing the binding this morning in my home installation of openHAB (arm Kubernetes cluster) and I'm getting a NotRouteToHostException when thing turns offline. Sorry for the late change and thanks for your help reviewing the pr.

**Review by cpmeister — APPROVED** [review]

**Comment by cpmeister** [comment]
Done
